### PR TITLE
gvox_palette: Improve `handle_single_palette` so it doesn't call `gvox_sample_region` twice for every voxel.

### PR DESCRIPTION
**Pull request by IntegratedQuantum**
In the given performance test this improved performance from 30ish to 22ish seconds for 1024³ voxels.

In general I just merged the two loops in `handle_single_palette`. To ensure that the behavior stays the same I freed `palette_region.data` at the end of t he function.
This comes at the cost of potentially needlessly allocating a bunch of memory, but I think that case should be rare enough to justify this. 

Additionally the code is now smaller.

### Future ideas
During benchmarking I discovered a few things that might be helpful in the future:
1. `gvox_sample_region` is quite costly, making up about 30-50% of the runtime (release). I think this could be reduced by allowing some sort of bulk access function, like maybe accessing 4³ regions at a time or so.
2. `std::unordered_set` uses a modulo based hash function. Modulo is quite costly and since we have to wait for the result, that one modulo alone takes around 15% of the entire runtime. This could be fixed by using a self made hash data structure that always uses power of two buffer sizes.
3. The memory usage goes quite high. Apparently `handle_single_palette` first copies the entire region into memory, but those regions stay in memory for quite a while, leading to 8 GB memory usage with the given test case.

### CODE REVIEW

**Comment by GabeRundlett** [comment]
I actually separated these two loops for the express purpose of not allocating for every 8x8x8 region. In the case of the 1024^3 set this is tested with, it allocates 8 bytes per voxel, and therefore 8GB of RAM. I would prefer if this wasn't the case.

What could be done is reverting the serialize-mode function `gvox_serialize_adapter_gvox_palette_serialize_region` to be more like what it was before I added the parse-driven mode. The only reason the temporary allocation happens in the first place is because it unifies/simplifies the code with the parse driven function `gvox_serialize_adapter_gvox_palette_receive_region`. If this were to happen, one would need to change `gvox_serialize_adapter_gvox_palette_blit_end` to also reflect what it was at that point in revision history (which is basically nothing). Another thing I noticed when looking at this code is that I appear to have forgotten to spawn job threads for `gvox_serialize_adapter_gvox_palette_blit_end`, meaning that function is completely serial at the moment

**Comment by GabeRundlett** [comment]
OH actually, yours doesn't allocate persistent memory. I missed the free at the end. This is great!
The reason it allocates that much is just because it appears the entire thing is non-uniform, I guess

File: src/adapters/serialize/gvox_palette.cpp [hunk 1]
```diff
@@ -257,43 +260,18 @@ static void handle_single_palette(
...
-                    auto const [prev_u32_voxel, prev_present] = (*palette_region.data)[palette_region_index];
-                    if (!prev_present && is_present) {
-                        (*palette_region.data)[palette_region_index] = {u32_voxel, is_present};
-                        ++palette_region.accounted_for;
-                    }
-                }
-            }
-        }
+    if(palette_region.accounted_for == 0) {
+        palette_region.data.reset(nullptr);
```
Comment (GabeRundlett) [inline src/adapters/serialize/gvox_palette.cpp]: Do you mind changing this to this? (space between the if and parenthesis, and removed `nullptr` from `.reset()`. The default overload is probably what I would use)
```cpp
if (palette_region.accounted_for == 0) {
    palette_region.data.reset();
}
```

**Review by GabeRundlett — CHANGES_REQUESTED** [review]
See my comment on line 274

File: src/adapters/serialize/gvox_palette.cpp (hunk 1 above)
Comment (IntegratedQuantum) [reply src/adapters/serialize/gvox_palette.cpp]: Done.

**Comment by GabeRundlett** [comment]
Thank you very much!

**Pull request merged**